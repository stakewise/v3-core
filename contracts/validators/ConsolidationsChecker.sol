// SPDX-License-Identifier: BUSL-1.1

pragma solidity ^0.8.22;

import {ECDSA} from '@openzeppelin/contracts/utils/cryptography/ECDSA.sol';
import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
import {IConsolidationsChecker} from '../interfaces/IConsolidationsChecker.sol';
import {Errors} from '../libraries/Errors.sol';
import {IKeeper} from '../interfaces/IKeeper.sol';

/**
 * @title ConsolidationsChecker
 * @author StakeWise
 * @notice Defines the functionality for checking signatures of oracles for validators consolidations
 */
contract ConsolidationsChecker is EIP712, IConsolidationsChecker {
  uint256 private constant _signatureLength = 65;
  bytes32 private constant _consolidationsCheckerTypeHash =
    keccak256('ConsolidationsChecker(address vault,bytes validators)');

  IKeeper private immutable _keeper;

  /**
   * @dev Constructor
   * @param keeper The address of the Keeper contract
   */
  constructor(address keeper) EIP712('ConsolidationsChecker', '1') {
    _keeper = IKeeper(keeper);
  }

  /// @inheritdoc IConsolidationsChecker
  function verifySignatures(
    address vault,
    bytes calldata validators,
    bytes calldata signatures
  ) external view override {
    if (!isValidSignatures(vault, validators, signatures)) {
      revert Errors.InvalidSignatures();
    }
  }

  /// @inheritdoc IConsolidationsChecker
  function isValidSignatures(
    address vault,
    bytes calldata validators,
    bytes calldata signatures
  ) public view override returns (bool) {
    return
      _isValidSignatures(
        _keeper.validatorsMinOracles(),
        keccak256(abi.encode(_consolidationsCheckerTypeHash, vault, keccak256(validators))),
        signatures
      );
  }

  /**
   * @notice Internal function for verifying oracles' signatures
   * @param requiredSignatures The number of signatures required for the verification to pass
   * @param message The message that was signed
   * @param signatures The concatenation of the oracles' signatures
   * @return True if the signatures are valid, otherwise false
   */
  function _isValidSignatures(
    uint256 requiredSignatures,
    bytes32 message,
    bytes calldata signatures
  ) private view returns (bool) {
    if (requiredSignatures == 0) {
      return false;
    }

    // check whether enough signatures
    unchecked {
      // cannot realistically overflow
      if (signatures.length < requiredSignatures * _signatureLength) {
        return false;
      }
    }

    bytes32 data = _hashTypedDataV4(message);
    address lastOracle;
    address currentOracle;
    uint256 startIndex;
    for (uint256 i = 0; i < requiredSignatures; i++) {
      unchecked {
        // cannot overflow as signatures.length is checked above
        currentOracle = ECDSA.recover(data, signatures[startIndex:startIndex + _signatureLength]);
      }
      // signatures must be sorted by oracles' addresses and not repeat
      if (currentOracle <= lastOracle || !_keeper.isOracle(currentOracle)) {
        return false;
      }

      // update last oracle
      lastOracle = currentOracle;

      unchecked {
        // cannot realistically overflow
        startIndex += _signatureLength;
      }
    }

    return true;
  }
}
