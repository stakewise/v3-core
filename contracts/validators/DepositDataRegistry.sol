// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.22;

import {MerkleProof} from '@openzeppelin/contracts/utils/cryptography/MerkleProof.sol';
import {IDepositDataRegistry} from '../interfaces/IDepositDataRegistry.sol';
import {IKeeperValidators} from '../interfaces/IKeeperValidators.sol';
import {IVaultAdmin} from '../interfaces/IVaultAdmin.sol';
import {IVaultValidators} from '../interfaces/IVaultValidators.sol';
import {IVaultVersion} from '../interfaces/IVaultVersion.sol';
import {IVaultsRegistry} from '../interfaces/IVaultsRegistry.sol';
import {IVaultState} from '../interfaces/IVaultState.sol';
import {IKeeperRewards} from '../interfaces/IKeeperRewards.sol';
import {Errors} from '../libraries/Errors.sol';
import {Multicall} from '../base/Multicall.sol';

/**
 * @title DepositDataRegistry
 * @author StakeWise
 * @notice Defines the functionality for the Vault's deposit data management
 */
contract DepositDataRegistry is Multicall, IDepositDataRegistry {
  IVaultsRegistry private immutable _vaultsRegistry;

  /// @inheritdoc IDepositDataRegistry
  mapping(address => uint256) public override depositDataIndexes;

  /// @inheritdoc IDepositDataRegistry
  mapping(address => bytes32) public override depositDataRoots;

  mapping(address => address) private _depositDataManagers;
  mapping(address => bool) private _migrated;

  /**
   * @dev Modifier to check if the caller is a valid vault
   * @param vault The address of the vault
   */
  modifier onlyValidVault(address vault) {
    if (!_vaultsRegistry.vaults(vault) || IVaultVersion(vault).version() < 2) {
      revert Errors.InvalidVault();
    }
    _;
  }

  /**
   * @dev Constructor
   * @param vaultsRegistry The address of the vaults registry contract
   */
  constructor(address vaultsRegistry) {
    _vaultsRegistry = IVaultsRegistry(vaultsRegistry);
  }

  /// @inheritdoc IDepositDataRegistry
  function getDepositDataManager(address vault) public view override returns (address) {
    address depositDataManager = _depositDataManagers[vault];
    return depositDataManager == address(0) ? IVaultAdmin(vault).admin() : depositDataManager;
  }

  /// @inheritdoc IDepositDataRegistry
  function setDepositDataManager(
    address vault,
    address depositDataManager
  ) external override onlyValidVault(vault) {
    // only vault admin can set deposit data manager
    if (msg.sender != IVaultAdmin(vault).admin()) revert Errors.AccessDenied();

    // update deposit data manager
    _depositDataManagers[vault] = depositDataManager;
    emit DepositDataManagerUpdated(vault, depositDataManager);
  }

  /// @inheritdoc IDepositDataRegistry
  function setDepositDataRoot(
    address vault,
    bytes32 depositDataRoot
  ) external override onlyValidVault(vault) {
    if (msg.sender != getDepositDataManager(vault)) revert Errors.AccessDenied();
    if (depositDataRoots[vault] == depositDataRoot) revert Errors.ValueNotChanged();

    depositDataRoots[vault] = depositDataRoot;
    // reset validator index on every root update
    depositDataIndexes[vault] = 0;
    emit DepositDataRootUpdated(vault, depositDataRoot);
  }

  /// @inheritdoc IDepositDataRegistry
  function updateVaultState(
    address vault,
    IKeeperRewards.HarvestParams calldata harvestParams
  ) external override {
    IVaultState(vault).updateState(harvestParams);
  }

  /// @inheritdoc IDepositDataRegistry
  function registerValidator(
    address vault,
    IKeeperValidators.ApprovalParams calldata keeperParams,
    bytes32[] calldata proof
  ) external override onlyValidVault(vault) {
    // register validator
    IVaultValidators(vault).registerValidators(keeperParams, '');

    // SLOAD to memory
    uint256 currentIndex = depositDataIndexes[vault];
    bytes32 depositDataRoot = depositDataRoots[vault];

    // check matches merkle root and next validator index
    if (
      !MerkleProof.verifyCalldata(
        proof,
        depositDataRoot,
        keccak256(bytes.concat(keccak256(abi.encode(keeperParams.validators, currentIndex))))
      )
    ) {
      revert Errors.InvalidProof();
    }

    // increment index for the next validator
    unchecked {
      // cannot realistically overflow
      depositDataIndexes[vault] = currentIndex + 1;
    }
  }

  /// @inheritdoc IDepositDataRegistry
  function registerValidators(
    address vault,
    IKeeperValidators.ApprovalParams calldata keeperParams,
    uint256[] calldata indexes,
    bool[] calldata proofFlags,
    bytes32[] calldata proof
  ) external override onlyValidVault(vault) {
    // register validator
    IVaultValidators(vault).registerValidators(keeperParams, '');

    // SLOAD to memory
    uint256 currentIndex = depositDataIndexes[vault];
    bytes32 depositDataRoot = depositDataRoots[vault];

    // define leaves for multiproof
    uint256 validatorsCount = indexes.length;
    if (validatorsCount == 0) revert Errors.InvalidValidators();
    bytes32[] memory leaves = new bytes32[](validatorsCount);

    // calculate validator length
    uint256 validatorLength = keeperParams.validators.length / validatorsCount;
    if (validatorLength == 0) revert Errors.InvalidValidators();

    // calculate leaves
    {
      uint256 startIndex;
      uint256 endIndex;
      for (uint256 i = 0; i < validatorsCount; ) {
        endIndex += validatorLength;
        leaves[indexes[i]] = keccak256(
          bytes.concat(
            keccak256(abi.encode(keeperParams.validators[startIndex:endIndex], currentIndex))
          )
        );

        startIndex = endIndex;
        unchecked {
          // cannot realistically overflow
          ++currentIndex;
          ++i;
        }
      }
    }

    // check matches merkle root and next validator index
    if (!MerkleProof.multiProofVerifyCalldata(proof, proofFlags, depositDataRoot, leaves)) {
      revert Errors.InvalidProof();
    }

    // increment index for the next validator
    depositDataIndexes[vault] = currentIndex;
  }

  /// @inheritdoc IDepositDataRegistry
  function migrate(
    bytes32 depositDataRoot,
    uint256 validatorIndex,
    address depositDataManager
  ) external override onlyValidVault(msg.sender) {
    if (_migrated[msg.sender]) revert Errors.AccessDenied();

    depositDataRoots[msg.sender] = depositDataRoot;
    depositDataIndexes[msg.sender] = validatorIndex;
    _depositDataManagers[msg.sender] = depositDataManager;

    // only allow migration once
    _migrated[msg.sender] = true;
    emit DepositDataMigrated(msg.sender, depositDataRoot, validatorIndex, depositDataManager);
  }
}
