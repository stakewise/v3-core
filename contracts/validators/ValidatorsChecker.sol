// SPDX-License-Identifier: BUSL-1.1

pragma solidity ^0.8.22;

import {MerkleProof} from '@openzeppelin/contracts/utils/cryptography/MerkleProof.sol';
import {MessageHashUtils} from '@openzeppelin/contracts/utils/cryptography/MessageHashUtils.sol';
import {SignatureChecker} from '@openzeppelin/contracts/utils/cryptography/SignatureChecker.sol';
import {IValidatorsRegistry} from '../interfaces/IValidatorsRegistry.sol';
import {IKeeper} from '../interfaces/IKeeper.sol';
import {IValidatorsChecker} from '../interfaces/IValidatorsChecker.sol';
import {IVaultState} from '../interfaces/IVaultState.sol';
import {IVaultVersion} from '../interfaces/IVaultVersion.sol';
import {IDepositDataRegistry} from '../interfaces/IDepositDataRegistry.sol';
import {IVaultsRegistry} from '../interfaces/IVaultsRegistry.sol';
import {IVaultValidators} from '../interfaces/IVaultValidators.sol';
import {Errors} from '../libraries/Errors.sol';

interface IVaultValidatorsV1 {
  function validatorsRoot() external view returns (bytes32);
  function validatorIndex() external view returns (uint256);
}

/**
 * @title ValidatorsChecker
 * @author StakeWise
 * @notice Defines the functionality for:
 *  * checking validators manager signature
 *  * checking deposit data root
 */
abstract contract ValidatorsChecker is IValidatorsChecker {
  bytes32 private constant _registerValidatorsTypeHash =
    keccak256('VaultValidators(bytes32 validatorsRegistryRoot,bytes validators)');

  IValidatorsRegistry private immutable _validatorsRegistry;
  IKeeper private immutable _keeper;
  IVaultsRegistry private immutable _vaultsRegistry;
  IDepositDataRegistry private immutable _depositDataRegistry;

  /**
   * @dev Constructor
   * @param validatorsRegistry The address of the beacon chain validators registry contract
   * @param keeper The address of the Keeper contract
   * @param vaultsRegistry The address of the VaultsRegistry contract
   * @param depositDataRegistry The address of the DepositDataRegistry contract
   */
  constructor(
    address validatorsRegistry,
    address keeper,
    address vaultsRegistry,
    address depositDataRegistry
  ) {
    _validatorsRegistry = IValidatorsRegistry(validatorsRegistry);
    _keeper = IKeeper(keeper);
    _vaultsRegistry = IVaultsRegistry(vaultsRegistry);
    _depositDataRegistry = IDepositDataRegistry(depositDataRegistry);
  }

  /// @inheritdoc IValidatorsChecker
  function checkValidatorsManagerSignature(
    address vault,
    bytes32 validatorsRegistryRoot,
    bytes calldata validators,
    bytes calldata signature
  ) external view override returns (uint256) {
    if (_validatorsRegistry.get_deposit_root() != validatorsRegistryRoot) {
      revert Errors.InvalidValidatorsRegistryRoot();
    }
    if (!_vaultsRegistry.vaults(vault) || IVaultVersion(vault).version() < 2) {
      revert Errors.InvalidVault();
    }

    // verify vault has enough assets
    if (
      !_keeper.isCollateralized(vault) && IVaultState(vault).withdrawableAssets() < _depositAmount()
    ) {
      revert Errors.InsufficientAssets();
    }

    // compose signing message
    bytes32 message = _getValidatorsManagerMessageHash(vault, validatorsRegistryRoot, validators);

    // verify validators manager ECDSA signature
    if (
      !SignatureChecker.isValidSignatureNow(
        IVaultValidators(vault).validatorsManager(),
        message,
        signature
      )
    ) {
      revert Errors.AccessDenied();
    }

    return block.number;
  }

  /// @inheritdoc IValidatorsChecker
  function checkDepositDataRoot(
    address vault,
    bytes32 validatorsRegistryRoot,
    bytes calldata validators,
    bytes32[] calldata proof,
    bool[] calldata proofFlags,
    uint256[] calldata proofIndexes
  ) external view override returns (uint256) {
    if (_validatorsRegistry.get_deposit_root() != validatorsRegistryRoot) {
      revert Errors.InvalidValidatorsRegistryRoot();
    }
    if (!_vaultsRegistry.vaults(vault)) revert Errors.InvalidVault();

    // verify vault has enough assets
    if (
      !_keeper.isCollateralized(vault) && IVaultState(vault).withdrawableAssets() < _depositAmount()
    ) {
      revert Errors.InsufficientAssets();
    }

    uint8 vaultVersion = IVaultVersion(vault).version();
    if (vaultVersion >= 2) {
      address validatorsManager = IVaultValidators(vault).validatorsManager();

      // verify vault did not set custom validators manager
      if (validatorsManager != address(_depositDataRegistry)) revert Errors.AccessDenied();
    }

    uint256 currentIndex;
    bytes32 depositDataRoot;

    if (vaultVersion >= 2) {
      currentIndex = _depositDataRegistry.depositDataIndexes(vault);
      depositDataRoot = _depositDataRegistry.depositDataRoots(vault);
    } else {
      currentIndex = IVaultValidatorsV1(vault).validatorIndex();
      depositDataRoot = IVaultValidatorsV1(vault).validatorsRoot();
    }

    // define leaves for multiproof
    uint256 validatorsCount = proofIndexes.length;
    if (validatorsCount == 0) revert Errors.InvalidValidators();
    bytes32[] memory leaves = new bytes32[](validatorsCount);

    // calculate validator length
    uint256 validatorLength = validators.length / validatorsCount;
    if (validatorLength == 0) revert Errors.InvalidValidators();

    // calculate leaves
    {
      uint256 startIndex;
      uint256 endIndex;
      for (uint256 i = 0; i < validatorsCount; ) {
        endIndex += validatorLength;
        leaves[proofIndexes[i]] = keccak256(
          bytes.concat(keccak256(abi.encode(validators[startIndex:endIndex], currentIndex)))
        );

        startIndex = endIndex;
        unchecked {
          // cannot realistically overflow
          ++currentIndex;
          ++i;
        }
      }
    }

    // check matches merkle root and next validator index
    if (!MerkleProof.multiProofVerifyCalldata(proof, proofFlags, depositDataRoot, leaves)) {
      revert Errors.InvalidProof();
    }

    return block.number;
  }

  /**
   * @notice Get the hash to be signed by the validators manager
   * @param vault The address of the vault
   * @param validatorsRegistryRoot The validators registry root
   * @param validators The concatenation of the validators' public key, deposit signature, deposit root and optionally withdrawal address
   * @return The hash to be signed by the validators manager
   */
  function _getValidatorsManagerMessageHash(
    address vault,
    bytes32 validatorsRegistryRoot,
    bytes calldata validators
  ) private view returns (bytes32) {
    bytes32 domainSeparator = _computeVaultValidatorsDomain(vault);
    return
      MessageHashUtils.toTypedDataHash(
        domainSeparator,
        keccak256(
          abi.encode(_registerValidatorsTypeHash, validatorsRegistryRoot, keccak256(validators))
        )
      );
  }

  /**
   * @notice Computes the hash of the EIP712 typed data for the vault
   * @dev This function is used to compute the hash of the EIP712 typed data
   * @return The hash of the EIP712 typed data
   */
  function _computeVaultValidatorsDomain(address vault) private view returns (bytes32) {
    return
      keccak256(
        abi.encode(
          keccak256(
            'EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)'
          ),
          keccak256(bytes('VaultValidators')),
          keccak256('1'),
          block.chainid,
          vault
        )
      );
  }

  /**
   * @notice Get the amount of assets required for validator deposit
   * @return The amount of assets required for deposit
   */
  function _depositAmount() internal pure virtual returns (uint256);
}
