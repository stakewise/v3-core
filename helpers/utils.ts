import { Contract } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/src/types/runtime'

export async function deployContract(tx: any): Promise<Contract> {
  const result = await tx
  await result.deployTransaction.wait()
  return result
}

async function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

export async function verify(
  hre: HardhatRuntimeEnvironment,
  address: string,
  constructorArgs: any,
  contractPath: string
) {
  if (!process.env.BLOCK_EXPLORER_KEY) {
    return
  }

  let count = 0
  const maxTries = 8
  // eslint-disable-next-line no-constant-condition
  while (true) {
    await delay(10000)
    try {
      console.log('Verifying contract at', address)
      await hre.run('verify:verify', {
        address,
        constructorArguments: constructorArgs,
        contract: contractPath,
      })
      break
    } catch (error) {
      if (String(error).includes('Already Verified')) {
        console.log(`Already verified contract at ${contractPath} at address ${address}`)
        break
      }
      if (++count == maxTries) {
        console.log(
          `Failed to verify contract at ${contractPath} at address ${address}, error: ${error}`
        )
        break
      }
      console.log(`Retrying... Retry #${count}, last error: ${error}`)
    }
  }
}
