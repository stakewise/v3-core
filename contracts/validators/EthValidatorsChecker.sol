// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.22;

import {ValidatorsChecker} from './ValidatorsChecker.sol';

/**
 * @title EthValidatorsChecker
 * @author StakeWise
 * @notice Defines functionality for checking validators registration on Ethereum
 */
contract EthValidatorsChecker is ValidatorsChecker {
  /**
   * @dev Constructor
   * @param validatorsRegistry The address of the beacon chain validators registry contract
   * @param keeper The address of the Keeper contract
   * @param vaultsRegistry The address of the VaultsRegistry contract
   * @param depositDataRegistry The address of the DepositDataRegistry contract
   */
  constructor(
    address validatorsRegistry,
    address keeper,
    address vaultsRegistry,
    address depositDataRegistry
  ) ValidatorsChecker(validatorsRegistry, keeper, vaultsRegistry, depositDataRegistry) {}

  /// @inheritdoc ValidatorsChecker
  function _depositAmount() internal pure override returns (uint256) {
    return 32 ether;
  }
}
