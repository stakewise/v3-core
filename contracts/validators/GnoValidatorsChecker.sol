// SPDX-License-Identifier: BUSL-1.1

pragma solidity =0.8.22;

import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
import {IValidatorsRegistry} from '../interfaces/IValidatorsRegistry.sol';
import {IKeeper} from '../interfaces/IKeeper.sol';
import {IDepositDataRegistry} from '../interfaces/IDepositDataRegistry.sol';
import {IVaultsRegistry} from '../interfaces/IVaultsRegistry.sol';
import {ValidatorsChecker} from './ValidatorsChecker.sol';

/**
 * @title GnoValidatorsChecker
 * @author StakeWise
 * @notice Defines Gnosis-specific settings for ValidatorsChecker contract
 */
contract GnoValidatorsChecker is ValidatorsChecker {
  /**
   * @dev Constructor
   * @param validatorsRegistry The address of the beacon chain validators registry contract
   * @param keeper The address of the Keeper contract
   * @param vaultsRegistry The address of the VaultsRegistry contract
   * @param depositDataRegistry The address of the DepositDataRegistry contract
   */
  constructor(
    address validatorsRegistry,
    address keeper,
    address vaultsRegistry,
    address depositDataRegistry
  )
    ValidatorsChecker(validatorsRegistry, keeper, vaultsRegistry, depositDataRegistry)
    EIP712('GnoValidatorsChecker', '1')
  {}

  function validatorsManagerSignatureTypeHash() internal pure override returns (bytes32) {
    return
      keccak256(
        'GnoValidatorsChecker(bytes32 validatorsRegistryRoot,address vault,bytes validators)'
      );
  }

  function depositAmount() internal pure override returns (uint256) {
    return 1 ether;
  }
}
