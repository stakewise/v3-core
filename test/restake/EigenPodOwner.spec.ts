import { loadFixture, mine } from '@nomicfoundation/hardhat-toolbox/network-helpers'
import { ethers } from 'hardhat'
import { Contract, parseEther, Signer, Wallet } from 'ethers'
import {
  EigenPodOwner,
  EigenPodOwner__factory,
  EthRestakeVault,
  EigenPodOwnerV2Mock__factory,
} from '../../typechain-types'
import { expect } from '../shared/expect'
import { ethRestakeVaultFixture } from '../shared/restakeFixtures'
import { MAX_UINT256, ZERO_ADDRESS, ZERO_BYTES32 } from '../shared/constants'
import { extractEigenPodOwner, setBalance } from '../shared/utils'
import {
  getEigenDelayedWithdrawalRouter,
  getEigenDelegationManager,
  getEigenPodManager,
} from '../shared/contracts'
import { registerEthValidator } from '../shared/validators'
import { MAINNET_FORK } from '../../helpers/constants'
import snapshotGasCost from '../shared/snapshotGasCost'

const gwei = 1000000000n
const validatorDeposit = parseEther('32')

describe('EigenPodOwner', () => {
  const capacity = MAX_UINT256
  const feePercent = 1000
  const metadataIpfsHash = 'bafkreidivzimqfqtoqxkrpge6bjyhlvxqs3rhe73owtmdulaxr5do5in7u'

  let admin: Signer, operatorsManager: Wallet, other: Wallet, withdrawalsManager: Wallet
  let vault: EthRestakeVault,
    eigenPodOwner: EigenPodOwner,
    delegationManager: Contract,
    eigenPodManager: Contract,
    delayedWithdrawalRouter: Contract
  let eigenPodAddress: string

  before('create fixture loader', async function () {
    ;[admin, operatorsManager, withdrawalsManager, other] = await (ethers as any).getSigners()
  })

  beforeEach('deploy fixture', async () => {
    const fixture = await loadFixture(ethRestakeVaultFixture)
    eigenPodManager = await getEigenPodManager()
    delegationManager = await getEigenDelegationManager()
    delayedWithdrawalRouter = await getEigenDelayedWithdrawalRouter()
    vault = await fixture.createEthRestakeVault(admin, {
      capacity,
      feePercent,
      metadataIpfsHash,
    })
    await vault.connect(admin).setRestakeOperatorsManager(operatorsManager.address)
    await vault.connect(admin).setRestakeWithdrawalsManager(withdrawalsManager.address)
    const receipt = await vault.connect(operatorsManager).createEigenPod()
    const eigenPodOwnerAddr = await extractEigenPodOwner(receipt)
    eigenPodOwner = EigenPodOwner__factory.connect(eigenPodOwnerAddr, admin)
    eigenPodAddress = (await vault.getEigenPods())[0]

    await vault.connect(other).deposit(other.address, ZERO_ADDRESS, { value: validatorDeposit })
    await registerEthValidator(
      vault,
      fixture.keeper,
      fixture.depositDataRegistry,
      admin,
      fixture.validatorsRegistry
    )
  })

  it('initializes correctly', async () => {
    await expect(eigenPodOwner.initialize(ZERO_BYTES32)).to.revertedWithCustomError(
      eigenPodOwner,
      'InvalidInitialization'
    )
    expect(await eigenPodOwner.vault()).to.equal(await vault.getAddress())
    const eigenPod = (await vault.getEigenPods())[0]
    expect(await eigenPodOwner.eigenPod()).to.equal(eigenPod)
  })

  describe('delegation', () => {
    beforeEach(async () => {
      const eigenPod = await ethers.getImpersonatedSigner(eigenPodAddress)
      await setBalance(eigenPod.address, parseEther('1'))
      await eigenPodManager
        .connect(eigenPod)
        .recordBeaconChainETHBalanceUpdate(eigenPodOwner, validatorDeposit / gwei)
    })

    it('not operators manager cannot delegate', async () => {
      await expect(
        eigenPodOwner
          .connect(other)
          .delegateTo(MAINNET_FORK.eigenOperator, { signature: '0x', expiry: 0 }, ZERO_BYTES32)
      ).to.revertedWithCustomError(eigenPodOwner, 'AccessDenied')
    })

    it('operators manager can delegate', async () => {
      const receipt = await eigenPodOwner
        .connect(operatorsManager)
        .delegateTo(MAINNET_FORK.eigenOperator, { signature: '0x', expiry: 0 }, ZERO_BYTES32)
      expect(await delegationManager.delegatedTo(await eigenPodOwner.getAddress())).to.be.eq(
        MAINNET_FORK.eigenOperator
      )
      await snapshotGasCost(receipt)
    })

    it('not operators manager cannot undelegate', async () => {
      await expect(eigenPodOwner.connect(other).undelegate()).to.revertedWithCustomError(
        eigenPodOwner,
        'AccessDenied'
      )
    })

    it('operator manager can undelegate', async () => {
      await eigenPodOwner
        .connect(operatorsManager)
        .delegateTo(MAINNET_FORK.eigenOperator, { signature: '0x', expiry: 0 }, ZERO_BYTES32)
      const receipt = await eigenPodOwner.connect(operatorsManager).undelegate()
      expect(await delegationManager.delegatedTo(await eigenPodOwner.getAddress())).to.be.eq(
        ZERO_ADDRESS
      )
      await snapshotGasCost(receipt)
    })
  })

  describe('withdrawals', () => {
    const withdrawalShares = validatorDeposit / gwei

    beforeEach(async () => {
      const eigenPod = await ethers.getImpersonatedSigner(eigenPodAddress)
      await setBalance(eigenPod.address, parseEther('1'))
      await eigenPodManager
        .connect(eigenPod)
        .recordBeaconChainETHBalanceUpdate(eigenPodOwner, withdrawalShares)
      await eigenPodOwner
        .connect(operatorsManager)
        .delegateTo(MAINNET_FORK.eigenOperator, { signature: '0x', expiry: 0 }, ZERO_BYTES32)
    })

    it('not withdrawals manager cannot queue withdrawal', async () => {
      await expect(
        eigenPodOwner.connect(other).queueWithdrawal(withdrawalShares)
      ).to.revertedWithCustomError(eigenPodOwner, 'AccessDenied')
    })

    it('withdrawals manager can queue withdrawal', async () => {
      const receipt = await eigenPodOwner
        .connect(withdrawalsManager)
        .queueWithdrawal(withdrawalShares / 2n)
      expect(await eigenPodManager.podOwnerShares(await eigenPodOwner.getAddress())).to.be.eq(
        withdrawalShares / 2n
      )
      await snapshotGasCost(receipt)
    })

    it('not withdrawals manager cannot complete withdrawal', async () => {
      await expect(
        eigenPodOwner
          .connect(other)
          .completeQueuedWithdrawal(MAINNET_FORK.eigenOperator, 0, withdrawalShares, 0, 0, false)
      ).to.revertedWithCustomError(eigenPodOwner, 'AccessDenied')
    })

    it('withdrawals manager can complete withdrawal', async () => {
      let receipt = await eigenPodOwner
        .connect(withdrawalsManager)
        .queueWithdrawal(withdrawalShares)
      const withdrawalBlockNumber = receipt.blockNumber as number
      const eigenPodOwnerAddress = await eigenPodOwner.getAddress()
      expect(await eigenPodManager.podOwnerShares(eigenPodOwnerAddress)).to.be.eq(0)

      await mine(await delegationManager.minWithdrawalDelayBlocks())
      receipt = await eigenPodOwner
        .connect(withdrawalsManager)
        .completeQueuedWithdrawal(
          MAINNET_FORK.eigenOperator,
          0,
          withdrawalShares,
          withdrawalBlockNumber,
          0,
          false
        )
      expect(await eigenPodManager.podOwnerShares(eigenPodOwnerAddress)).to.be.eq(withdrawalShares)
      await snapshotGasCost(receipt)
    })

    it('not eigen pod or delayed withdrawals manager cannot transfer assets', async () => {
      await expect(
        other.sendTransaction({
          to: await eigenPodOwner.getAddress(),
          value: parseEther('1'),
        })
      ).to.be.revertedWithCustomError(eigenPodOwner, 'AccessDenied')
    })

    it('can claim delayed withdrawals', async () => {
      const eigenPod = await ethers.getImpersonatedSigner(eigenPodAddress)
      const eigenPodOwnerAddress = await eigenPodOwner.getAddress()
      await setBalance(eigenPod.address, parseEther('11'))

      const withdrawalAmount = parseEther('10')
      await delayedWithdrawalRouter
        .connect(eigenPod)
        .createDelayedWithdrawal(eigenPodOwnerAddress, eigenPodOwnerAddress, {
          value: withdrawalAmount,
        })

      const totalAssetsBefore = await vault.totalAssets()
      const totalSharesBefore = await vault.totalShares()
      const vaultBalanceBefore = await ethers.provider.getBalance(await vault.getAddress())

      await mine(await delayedWithdrawalRouter.withdrawalDelayBlocks())
      const receipt = await eigenPodOwner.claimDelayedWithdrawals(1)

      expect(await vault.totalAssets()).to.be.eq(totalAssetsBefore)
      expect(await vault.totalShares()).to.be.eq(totalSharesBefore)
      expect(await ethers.provider.getBalance(await vault.getAddress())).to.be.eq(
        vaultBalanceBefore + withdrawalAmount
      )

      await snapshotGasCost(receipt)
    })
  })
  describe('upgrade', () => {
    it('fails to upgrade by not vault', async () => {
      await expect(
        eigenPodOwner.connect(other).upgradeToAndCall(other.address, '0x')
      ).to.revertedWithCustomError(eigenPodOwner, 'AccessDenied')
    })

    it('fails to upgrade to same implementation', async () => {
      const vaultSigner = await ethers.getImpersonatedSigner(await vault.getAddress())
      await setBalance(vaultSigner.address, parseEther('1'))

      await expect(
        eigenPodOwner
          .connect(vaultSigner)
          .upgradeToAndCall(await eigenPodOwner.implementation(), '0x')
      ).to.revertedWithCustomError(eigenPodOwner, 'UpgradeFailed')
    })

    it('fails to upgrade to zero address', async () => {
      const vaultSigner = await ethers.getImpersonatedSigner(await vault.getAddress())
      await setBalance(vaultSigner.address, parseEther('1'))
      await expect(
        eigenPodOwner.connect(vaultSigner).upgradeToAndCall(ZERO_ADDRESS, '0x')
      ).to.revertedWithCustomError(eigenPodOwner, 'UpgradeFailed')
    })

    it('succeeds to upgrade', async () => {
      const vaultSigner = await ethers.getImpersonatedSigner(await vault.getAddress())
      await setBalance(vaultSigner.address, parseEther('1'))

      const newImplementation = await ethers.getContractFactory('EigenPodOwnerV2Mock')
      const newImplementationAddress = await newImplementation.deploy(
        MAINNET_FORK.eigenPodManager,
        MAINNET_FORK.eigenDelegationManager,
        MAINNET_FORK.eigenDelayedWithdrawalRouter
      )
      const receipt = await eigenPodOwner
        .connect(vaultSigner)
        .upgradeToAndCall(newImplementationAddress, '0x')

      const eigenPodOwnerV2 = EigenPodOwnerV2Mock__factory.connect(
        await eigenPodOwner.getAddress(),
        admin
      )
      expect(await eigenPodOwnerV2.implementation()).to.be.eq(newImplementationAddress)
      expect(await eigenPodOwnerV2.somethingNew()).to.be.eq(true)
      await snapshotGasCost(receipt)
    })
  })
})
