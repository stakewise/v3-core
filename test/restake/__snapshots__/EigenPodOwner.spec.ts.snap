// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EigenPodOwner delegation operator manager can undelegate 1`] = `
Object {
  "calldataByteLength": 4,
  "gasUsed": 146611,
}
`;

exports[`EigenPodOwner delegation operators manager can delegate 1`] = `
Object {
  "calldataByteLength": 196,
  "gasUsed": 118993,
}
`;

exports[`EigenPodOwner upgrade succeeds to upgrade 1`] = `
Object {
  "calldataByteLength": 100,
  "gasUsed": 37475,
}
`;

exports[`EigenPodOwner withdrawals can claim delayed withdrawals 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 98777,
}
`;

exports[`EigenPodOwner withdrawals withdrawals manager can complete withdrawal 1`] = `
Object {
  "calldataByteLength": 196,
  "gasUsed": 115980,
}
`;

exports[`EigenPodOwner withdrawals withdrawals manager can queue withdrawal 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 136023,
}
`;
